File: backend/app/api/assignments.py
from datetime import datetime

from fastapi import APIRouter, Depends, File, Form, Query, UploadFile
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.responses import success
from app.db.session import get_db
from app.models.user import User
from app.schemas.assignment_schema import AssignmentCreate, AssignmentUpdate
from app.services import assignment_service

router = APIRouter(prefix="/assignments", tags=["作业"])


@router.get("")
def list_assignments(
    course_id: int | None = Query(default=None),
    only_mine: bool = Query(default=False),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    data = assignment_service.list_assignments(db, user, course_id=course_id, only_mine=only_mine)
    return success([item.model_dump() for item in data])


@router.post("")
def create_assignment(payload: AssignmentCreate, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    return success(assignment_service.create_assignment(db, user, payload).model_dump())


@router.post("/with-files")
def create_assignment_with_files(
    course_id: int = Form(...),
    title: str = Form(...),
    deadline: datetime = Form(...),
    description: str | None = Form(default=None),
    files: list[UploadFile] | None = File(default=None),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    payload = AssignmentCreate(course_id=course_id, title=title, description=description, deadline=deadline)
    return success(assignment_service.create_assignment_with_files(db, user, payload, files).model_dump())


@router.get("/{assignment_id}")
def get_assignment(assignment_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    return success(assignment_service.get_assignment(db, user, assignment_id).model_dump())


@router.put("/{assignment_id}")
def update_assignment(
    assignment_id: int,
    payload: AssignmentUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    return success(assignment_service.update_assignment(db, user, assignment_id, payload).model_dump())


@router.delete("/{assignment_id}")
def delete_assignment(assignment_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    assignment_service.delete_assignment(db, user, assignment_id)
    return success(None, message="删除成功")

File: frontend/src/api/assignments.ts
import request from './http'
import type { Assignment, AssignmentPayload } from './types'

export function getAssignments(params?: { course_id?: number; only_mine?: boolean }) {
  return request.get<unknown, Assignment[]>('/assignments', { params })
}

export function getAssignment(id: number) {
  return request.get<unknown, Assignment>(`/assignments/${id}`)
}

export function createAssignment(payload: AssignmentPayload) {
  return request.post<unknown, Assignment>('/assignments', payload)
}

export function createAssignmentWithFiles(payload: AssignmentPayload, files: File[]) {
  const data = new FormData()
  data.append('course_id', String(payload.course_id))
  data.append('title', payload.title)
  data.append('deadline', payload.deadline)
  data.append('description', payload.description || '')
  files.forEach((file) => data.append('files', file))
  return request.post<unknown, Assignment>('/assignments/with-files', data)
}

export function updateAssignment(id: number, payload: Omit<AssignmentPayload, 'course_id'>) {
  return request.put<unknown, Assignment>(`/assignments/${id}`, payload)
}

export function deleteAssignment(id: number) {
  return request.delete<unknown, null>(`/assignments/${id}`)
}
